pkg/internal/controllers/test/util.go

```diff
@@ -122,48 +122,59 @@ func waitForNotReady(ctx context.Context, cl client.Client, name string) {
 // created namespace as well as a shutdown function to stop the controllers is
 // returned.
 func startControllers(registry *registry.Registry) (context.Context, func(), corev1.Namespace) {
-	ctx, cancel := context.WithCancel(context.Background())
+	log, ctx := ktesting.NewTestContext(GinkgoT())
+	ctx, cancel := context.WithCancel(ctx)
 
 	namespace := corev1.Namespace{
 		ObjectMeta: metav1.ObjectMeta{
 			GenerateName: "test-policy-",
 		},
 	}
-	Expect(env.AdminClient.Create(ctx, &namespace)).NotTo(HaveOccurred())
 	By("Created Policy Namespace: " + namespace.Name)
+	Expect(env.AdminClient.Create(ctx, &namespace)).NotTo(HaveOccurred())
+
+	mgrStopped := make(chan struct{})
 
 	shutdown := func() {
+		cancel()
+		<-mgrStopped
+
 		By("Deleting all CertificateRequests after test")
-		env.AdminClient.DeleteAllOf(ctx, new(cmapi.CertificateRequest))
+		Expect(
+			client.IgnoreNotFound(
+				env.AdminClient.DeleteAllOf(context.TODO(), new(cmapi.CertificateRequest)),
+			),
+		).To(Succeed())
 
 		By("Deleting all CertificateRequestPolicies after test")
-		env.AdminClient.DeleteAllOf(ctx, new(policyapi.CertificateRequestPolicy))
+		Expect(env.AdminClient.DeleteAllOf(context.TODO(), new(policyapi.CertificateRequestPolicy))).To(Succeed())
 
 		By("Deleting Policy Namespace: " + namespace.Name)
-		env.AdminClient.Delete(ctx, &namespace)
-
-		cancel()
+		Expect(env.AdminClient.Delete(context.TODO(), &namespace)).To(Succeed())
 	}
-
-	log := klogr.New().WithName("testing")
 	mgr, err := ctrl.NewManager(env.Config, ctrl.Options{
 		Scheme:                        policyapi.GlobalScheme,
 		LeaderElection:                true,
 		MetricsBindAddress:            "0",
-		LeaderElectionNamespace:       namespace.Name,
+		LeaderElectionNamespace:       "default",
```

Comment:
can we run all tests in parallel (each reconciling resources in their individual namespace)?

Reply:
No, because the CertificateRequestPolicy is cluster-scoped, so the parallel controllers would end up reconciling policies from other tests.

And Ginkgo doesn't do goroutine based parallel tests anyway.
It uses sub-processes I think.
*  https://onsi.github.io/ginkgo/#spec-parallelization
So each process would have it's own dedicated envtest etcd and kube-apiserver.